### DISCUSSION TITLE
Bluedroid and NimBLE stack simultaneously

### DISCUSSION THREAD

**Discussion by cyberFIRM** [h2zero/NimBLE-Arduino#634] [Q&A] [answered]
Has anyone ever tried to implement BLE beacon and BT classic in single sketch and running both simultaneously in dual mode?

Framework: Arduino
Dev Kit: ESP32 DEVkit 1
Platform: PlatformIO core 6.1.13

### Application example:
I have a positioning proximity beacon. I want ESP32 to check if someone is within certain distance (lets say 1 meter), ESP32 plays an audio on the Bluetooth speaker.

I am using A2DP/Audio tools by @pschatzmann and using player-sdfat-a2dp example as a reference. I have tried to implement this application and got following error when connected to bluetooth speaker:

```
[I] AudioA2DP.h : 126 - Connecting to E600Pro
[I] AudioA2DP.h : 137 - Starting a2dp_source...
ESP_ERROR_CHECK failed: esp_err_t 0x103 (ESP_ERR_INVALID_STATE) at 0x400944a0
file: ".pio/libdeps/esp32dev/ESP32-A2DP/src/BluetoothA2DPSource.cpp" line 203
func: virtual void BluetoothA2DPSource::start_raw(std::vector<const char*>, music_data_cb_t)
expression: esp_bt_controller_mem_release(ESP_BT_MODE_BLE)

abort() was called at PC 0x400944a3 on core 1
```

### Things I tried:

Identified the problem that esp_bt_controller_enable [can't be called twice using different configs](https://github.com/espressif/esp-idf/blob/165e3edd4949586dc90fa682f907d971fee488fc/components/bt/include/esp_bt.h#L161) and both libraries call it when they start
For both work together, they should call esp_bt_controller_enable(ESP_BT_MODE_BTDM) once, and not the BLE lib call esp_bt_controller_enable(ESP_BT_MODE_BLE) and BluetoohSerial call esp_bt_controller_enable(ESP_BT_MODE_CLASSIC_BT)

By doing this, I was able to connect to Bluetooth speaker but could not play any audio. For testing and debugging purposes,  I went with BluetoothSerial library and begin BT in slave mode. The logs below showcase ESP32 finds the BLE beacon and see if it is within 1 meter range. If yes it stops scanning, then it begin BT in slave mode. On pairing ESP32 with my phone, I sent a string "hello" to ESP32. On receiving, ESP32 should resume the scan again But what I see it that BLE host gets reset and after that I get GAP init failed error.

```
rst:0x1 (POWERON_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0030,len:1184
load:0x40078000,len:13192
load:0x40080400,len:3028
entry 0x400805e4
E (599) esp_core_dump_flash: No core dump  …ÉÑ¥Ñ¥½¹�found!
E (599) esp_core_dump_flash: No core dump partition found!
firmware version 0.4
1
Scanning for peripherals
Advertised Device found: Name: , Address: 57:04:4b:a3:24:6b, manufacturer data: 4c001005311cb7c9f9, txPower: 12
Advertised Device found: Name: , Address: 5a:48:e8:ea:a7:b1, manufacturer data: 060001092002cc4e59b1f9d7c6284db5f38bca3482078467229cbc024e
Advertised Device found: Name: , Address: 49:ca:64:9f:8d:3c, manufacturer data: 4c0010021300, txPower: 12
Advertised Device found: Name: , Address: e7:be:62:3a:61:dc, manufacturer data: 4c0012020000
Found Beacon
-82
@1m
Scan Ended
HERE !!
BT init success !!
hello
CRIT NimBLEDevice: Resetting state; reason=19, 
CRIT NimBLEScan: Unable to scan - Host Reset
0
Scanning for peripherals
E NimBLEScan: Error initiating GAP discovery procedure; rc=21, 
0
```

Any ideas?

Thanks

**Comment by h2zero** [comment, marked as answer]
For this I would just use the bluedroid library as provided by the core. By using nimble you will be introducing a lot of complexity for little gain.

**Reply by cyberFIRM** [reply]
Thanks. Resolved
